demo/demo04_basic_attacker.py

```diff
@@ -0,0 +1,72 @@
+# This bot selects a food pellet at random, then goes and tries to get it by
+# following the shortest path to it.
+# It tries on the way to avoid being eaten by the enemy: if the next move
+# to get to the food would put it on a ghost, then it chooses a random safe
+# position
+TEAM_NAME = 'Basic Attacker Bots'
+
+from pelita.utils import Graph
+
+from utils import shortest_path
+
+
+def move(bot, state):
+    enemy = bot.enemy
+    # we need to create a dictionary to keep information (state) along rounds
+    # the state object will be passed untouched at every new round
+    if state is None:
+        # initialize the state dictionary
+        state = {}
+        # each bot needs its own state dictionary to keep track of the
+        # food targets
+        state[0] = (None, None)
+        state[1] = (None, None)
+        # initialize a graph representation of the maze
+        # this can be shared among our bots
+        state['graph'] = Graph(bot.position, bot.walls)
+
+    target, path = state[bot.turn]
+
+    # choose a target food pellet if we still don't have one or
+    # if the old target has been already eaten
+    if (target is None) or (target not in enemy[0].food):
+        # position of the target food pellet
+        target = bot.random.choice(enemy[0].food)
+        # shortest path from here to the target
+        path = shortest_path(bot.position, target, state['graph'])
+        state[bot.turn] = (target, path)
+
+    # get the next position along the shortest path to reach our target
+    next_pos = path.pop()
+    # get a list of safe positions
+    safe_positions = []
+    for pos in bot.legal_positions:
+        # a position is safe if the enemy is not sitting on it *and*
+        # the enemy does not sit in the neighborhood of that position
+        safe = True
+        for direction in ((0, 0), (0,1), (0,-1), (1,0), (-1,0)):
```

Comment:
I am undecided with this. The more helper functions we have, the more we have to explain. This one is pretty easy to code yourself and if you make mistakes (like forgetting `(0,0)` like we did in the first iteration of this function) it is actually a good thing: you learn by doing mistakes.

But this demo is only there because we figured that this bot is too good to be given in `pelita_template`: it successfully escapes single ghosts 100% of the times. I put it there just not to lose the code, but it should probably go away soon and land in some "hidden" repo of good bots, to be used for the network-bot offer during the tournament.

Reply:
Yeah, I agree. (Did not test the bot to see how good it is.)